File: lib/python/opentoken/src/main/opentoken/keyexchange/public_key_loader.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import os
import logging
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.backends import default_backend

from opentoken.keyexchange.key_exchange_exception import KeyExchangeException


logger = logging.getLogger(__name__)


class PublicKeyLoader:
    """
    Loads and validates public keys for OpenToken key exchange.

    Supports loading EC public keys in PEM format from files or strings,
    and validates that they use the expected curve (P-256/secp256r1).
    """

    def load_public_key(self, file_path: str) -> ec.EllipticCurvePublicKey:
        """
        Loads a public key from a PEM file.

        Args:
            file_path: The path to the PEM file containing the public key.

        Returns:
            The loaded public key.

        Raises:
            KeyExchangeException: If loading or validation fails.
        """
        try:
            logger.debug(f"Loading public key from {file_path}")

            # Check if file exists
            if not os.path.exists(file_path):
                raise KeyExchangeException(f"Public key file not found: {file_path}")

            with open(file_path, 'rb') as f:
                public_key = serialization.load_pem_public_key(
                    f.read(),
                    backend=default_backend()
                )

            self.validate_public_key(public_key)

            logger.info(f"Successfully loaded and validated public key from {file_path}")
            return public_key
        except Exception as e:
            if isinstance(e, KeyExchangeException):
                raise
            raise KeyExchangeException(f"Failed to load public key from {file_path}", e)

    def load_public_key_from_string(self, pem_content: str) -> ec.EllipticCurvePublicKey:
        """
        Loads a public key from a PEM-encoded string.

        Args:
            pem_content: The PEM-encoded public key content.

        Returns:
            The loaded public key.

        Raises:
            KeyExchangeException: If loading or validation fails.
        """
        try:
            logger.debug("Loading public key from PEM string")

            # Convert string to bytes if needed
            if isinstance(pem_content, str):
                pem_content = pem_content.encode('utf-8')

            public_key = serialization.load_pem_public_key(
                pem_content,
                backend=default_backend()
            )

            self.validate_public_key(public_key)

            logger.info("Successfully loaded and validated public key from PEM string")
            return public_key
        except Exception as e:
            if isinstance(e, KeyExchangeException):
                raise
            raise KeyExchangeException("Failed to load public key from PEM string", e)

    def validate_public_key(self, public_key) -> None:
        """
        Validates that a public key is an EC key using the expected curve.

        Args:
            public_key: The public key to validate.

        Raises:
            KeyExchangeException: If validation fails.
        """
        if public_key is None:
            raise KeyExchangeException("Public key is None")

        # Check if it's an EC key
        if not isinstance(public_key, ec.EllipticCurvePublicKey):
            raise KeyExchangeException(
                f"Invalid key type: expected EC public key, got {type(public_key).__name__}"
            )

        # Get the curve
        curve = public_key.curve
        curve_name = self._get_curve_name(curve)
        logger.debug(f"Public key uses curve: {curve_name}")

        # Validate it's a supported curve (P-256)
        if not isinstance(curve, ec.SECP256R1):
            logger.warning(f"Public key uses curve {curve_name}, expected SECP256R1 (P-256)")

        logger.debug("Public key validation passed")

    def public_key_file_exists(self, file_path: str) -> bool:
        """
        Checks if a public key file exists.

        Args:
            file_path: The path to check.

        Returns:
            True if the file exists, False otherwise.
        """
        return os.path.exists(file_path)

    def _get_curve_name(self, curve: ec.EllipticCurve) -> str:
        """
        Gets the curve name from an EC curve object.

        Args:
            curve: The EC curve.

        Returns:
            A string representing the curve name.
        """
        if isinstance(curve, ec.SECP256R1):
            return "P-256/secp256r1"
        elif isinstance(curve, ec.SECP384R1):
            return "P-384/secp384r1"
        elif isinstance(curve, ec.SECP521R1):
            return "P-521/secp521r1"
        else:
            return f"Unknown ({curve.name})"

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/keyexchange/PublicKeyLoader.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.keyexchange;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.NoSuchAlgorithmException;
import java.security.PublicKey;
import java.security.interfaces.ECPublicKey;
import java.security.spec.InvalidKeySpecException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Loads and validates public keys for OpenToken key exchange.
 * <p>
 * Supports loading EC public keys in PEM format from files or strings,
 * and validates that they use the expected curve (P-256/secp256r1).
 */
public class PublicKeyLoader {
    private static final Logger logger = LoggerFactory.getLogger(PublicKeyLoader.class);
    
    /**
     * Loads a public key from a PEM file.
     *
     * @param filePath the path to the PEM file containing the public key
     * @return the loaded public key
     * @throws KeyExchangeException if loading or validation fails
     */
    public PublicKey loadPublicKey(String filePath) throws KeyExchangeException {
        try {
            logger.debug("Loading public key from {}", filePath);
            
            // Check if file exists
            if (!Files.exists(Paths.get(filePath))) {
                throw new KeyExchangeException("Public key file not found: " + filePath);
            }
            
            PublicKey publicKey = PemUtils.loadPublicKey(filePath);
            validatePublicKey(publicKey);
            
            logger.info("Successfully loaded and validated public key from {}", filePath);
            return publicKey;
        } catch (IOException | NoSuchAlgorithmException | InvalidKeySpecException e) {
            throw new KeyExchangeException("Failed to load public key from " + filePath, e);
        }
    }
    
    /**
     * Loads a public key from a PEM-encoded string.
     *
     * @param pemContent the PEM-encoded public key content
     * @return the loaded public key
     * @throws KeyExchangeException if loading or validation fails
     */
    public PublicKey loadPublicKeyFromString(String pemContent) throws KeyExchangeException {
        try {
            logger.debug("Loading public key from PEM string");
            
            PublicKey publicKey = PemUtils.loadPublicKeyFromString(pemContent);
            validatePublicKey(publicKey);
            
            logger.info("Successfully loaded and validated public key from PEM string");
            return publicKey;
        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
            throw new KeyExchangeException("Failed to load public key from PEM string", e);
        }
    }
    
    /**
     * Validates that a public key is an EC key using the expected curve.
     *
     * @param publicKey the public key to validate
     * @throws KeyExchangeException if validation fails
     */
    public void validatePublicKey(PublicKey publicKey) throws KeyExchangeException {
        if (publicKey == null) {
            throw new KeyExchangeException("Public key is null");
        }
        
        // Check if it's an EC key
        if (!(publicKey instanceof ECPublicKey)) {
            throw new KeyExchangeException(
                "Invalid key type: expected EC public key, got " + publicKey.getClass().getName()
            );
        }
        
        ECPublicKey ecPublicKey = (ECPublicKey) publicKey;
        
        // Get the curve name (this is a simplified check)
        String curveName = getCurveName(ecPublicKey);
        logger.debug("Public key uses curve: {}", curveName);
        
        // Validate the algorithm
        if (!KeyPairManager.EC_ALGORITHM.equals(publicKey.getAlgorithm())) {
            throw new KeyExchangeException(
                "Invalid key algorithm: expected " + KeyPairManager.EC_ALGORITHM + 
                ", got " + publicKey.getAlgorithm()
            );
        }
        
        // Note: More rigorous curve validation could be added here
        // For now, we trust that if it's an EC key, it's using a standard curve
        logger.debug("Public key validation passed");
    }
    
    /**
     * Gets the curve name from an EC public key.
     * <p>
     * This is a best-effort method that returns a string representation
     * of the curve parameters.
     *
     * @param ecPublicKey the EC public key
     * @return a string representing the curve (simplified)
     */
    private String getCurveName(ECPublicKey ecPublicKey) {
        // Get curve parameters
        java.security.spec.ECParameterSpec params = ecPublicKey.getParams();
        
        // Get the curve field size (bit length)
        int fieldSize = params.getCurve().getField().getFieldSize();
        
        // Map field size to common curve names
        switch (fieldSize) {
            case 256:
                return "P-256/secp256r1";
            case 384:
                return "P-384/secp384r1";
            case 521:
                return "P-521/secp521r1";
            default:
                return "Unknown (" + fieldSize + " bits)";
        }
    }
    
    /**
     * Checks if a public key file exists.
     *
     * @param filePath the path to check
     * @return true if the file exists, false otherwise
     */
    public boolean publicKeyFileExists(String filePath) {
        return Files.exists(Paths.get(filePath));
    }
}
